File: wc-clone/wc-javascript/wc.test.js
import fs from 'node:fs';

import { jest } from '@jest/globals';

import { Readable } from 'stream';

import { runWordCount, countBytes, countWords, countLines, countChars } from './wc.js';

describe('word count functions', () => {
    afterEach(() => {
        // clean up the temp file after each test
        if (fs.existsSync('test.txt')) {
            fs.unlinkSync('test.txt');
        } 
    });

    test('counts the number of lines in a file', () => {
        fs.writeFileSync('test.txt', 'Line 1\nLine 2\nLine 3\n');

        const count = countLines('test.txt');

        expect(count).toBe(3);
    });

    test('counts lines, words, and bytes when no option is provided', async () => {
        fs.writeFileSync('test.txt', 'Line 1\nLine 2\nLine 3\n');

        const result = await runWordCount(['node', 'wc.js', 'test.txt']);

        expect(result).toBe('3 6 21 test.txt')
    });

    test('counts the number of bytes in a file', () => {
        // Create a temp file with sample content
        fs.writeFileSync('test.txt', 'Sample content');

        // Call the countsBytes function
        const count = countBytes('test.txt');

        // Assert the expected byte count
        expect(count).toBe(14);
    }); 

    test('returns null for a non-existent file', () => {
        // Call the countBytes function with a non-existent file
        const count = countBytes('nonexistent.txt');

        // Assert that the count is null (indicating an error) 
        expect(count).toBeNull();
    });

    test('counts the number of words in a file', () => {
        fs.writeFileSync('test.txt', 'This is a sample file\nwith multiple words\non each line\n');

        const count = countWords('test.txt');

        expect(count).toBe(11);
    }); 

    test('counts the number of characters in a file', () => {
        fs.writeFileSync('test.txt', 'Sample content with 🚀 emoji');

        const charCount = countChars('test.txt');
        const byteCount = countBytes('test.txt');

        expect(charCount).toBe(27);
        expect(charCount).not.toBe(byteCount);
    }); 

    test('reads from standard input when no filename is provided', async () => {
        const stdin = 'Line 1\nLine 2\nLine 3\n';
        const expectedOutput = '3 6 21';

        const stdinStream = new Readable();
        stdinStream.push(stdin);
        stdinStream.push(null);

        // Mock standard input
        const stdinMock = jest.spyOn(process, 'stdin', 'get').mockReturnValue(stdinStream);

        // capture the console output
        const consoleLogMock = jest.spyOn(console, 'log').mockImplementation();

        await runWordCount(['node', 'wc.js']);

        expect(consoleLogMock).toHaveBeenCalledWith(expectedOutput);

        stdinMock.mockRestore();
        consoleLogMock.mockRestore()
    });
});

File: wc-clone/wc-elixir/test/wc_test.exs
defmodule WcTest do 
  use ExUnit.Case
  import ExUnit.CaptureIO
  doctest Wc

  test "main/1 counts lines, words, and bytes when no option is provided" do 
    content = "Line 1\nLine 2\nLine 3\n"
    file_name = "test.txt"
    File.write!(file_name, content)
    expected_output = "3 6 21 test.txt\n"
    assert capture_io(fn -> Wc.main([file_name]) end) == expected_output
    File.rm!(file_name)
  end

  test "count_bytes/1 counts the number of bytes in a file" do 
    content = "Sample content"
    file_name = "test.txt"
    File.write!(file_name, content)
    assert Wc.count_bytes(file_name) == {:ok, byte_size(content)}
    File.rm!(file_name)
  end

  test "count_bytes/1 returns an error for a non-existent file" do
    file_name = "nonexistent.txt"
    assert Wc.count_bytes(file_name) == {:error, :file_not_found}
  end

  test "count_lines/1 counts the number of lines in a file" do 
    content = "Line 1\nLine 2\nLine 3\n"
    file_name = "test.txt"
    File.write!(file_name, content)
    assert Wc.count_lines(file_name) == {:ok, 3}
    File.rm!(file_name)
  end

  test "count_lines/1 returns an error for a non-existent file" do 
    file_name = "nonexistent.txt"
    assert Wc.count_lines(file_name) == {:error, :file_not_found}
  end

  test "count_words/1 counts the number of words in a file" do 
    content = "This is a sample file\nwith multiple words\non each line\n"
    file_name = "test.txt"
    File.write!(file_name, content)
    assert Wc.count_words(file_name) == {:ok, 11}
    File.rm!(file_name)
  end

  test "count_words/1 returns an error for a non-existent file" do 
    file_name = "nonexistent.txt"
    assert Wc.count_words(file_name) == {:error, :file_not_found}
  end

  test "count_chars/1 counts the number of characters in a file" do 
    content = "Sample content with 🚀 emoji"
    file_name = "test.txt"
    File.write!(file_name, content)
    assert Wc.count_chars(file_name) == {:ok, String.length(content)}
    File.rm!(file_name)
  end

  test "count_chars/1 returns an error for a non-existent file" do 
    file_name = "nonexistent.txt"
    assert Wc.count_chars(file_name) == {:error, :file_not_found}
  end

end

